carla/recourse_methods/catalog/face/library/face_method.py

```diff
@@ -0,0 +1,282 @@
+# import helpers
+import numpy as np
+
+# import Dijkstra's shortest path algorithm
+from scipy.sparse import csgraph, csr_matrix
+
+# import graph building methods
+from sklearn.neighbors import kneighbors_graph, radius_neighbors_graph
+
+
+def graph_search(
+    data,
+    index,
+    keys_immutable,
+    model,
+    n_neighbors=50,
+    p_norm=2,
+    mode="knn",
+    frac=0.4,
+    radius=0.25,
+):
+    # This one implements the FACE method from
+    # Rafael Poyiadzi et al (2020), "FACE: Feasible and Actionable Counterfactual Explanations",
+    # Conference on AI, Ethics & Accountability (AIES, 2020)
+    """
+    :param data: df
+    :param n_neighbors: int > 0; number of neighbors when constructing knn graph
+    :param step: float > 0; step_size for growing spheres
+    :param mode: str; either 'knn' or 'epsilon'
+    :param model: classification model (either tf keras, pytorch or sklearn)
+    :param p_norm: float=>1; denotes the norm (classical: 1 or 2)
+    :param frac: float 0 < number =< 1; fraction of data for which we compute the graph; if frac = 1, and data set large, then compute long
+    :param keys_immutable: list; list of input names that may not be searched over
+    :param radius: float > 0; parameter for epsilon density graph
+    :return: candidate_counterfactual_star: np array (min. cost counterfactual explanation)
+    """
+    # Choose a subset of data for computational efficiency
+    data = choose_random_subset(data, frac, index)
+
+    # ADD CONSTRAINTS by immutable inputs into adjacency matrix
+    # if element in adjacency matrix 0, then it cannot be reached
+    # this ensures that paths only take same sex / same race / ... etc. routes
+    # TODO: CANNOT deal with continuous constraints (e.g. age) yet
+    # TODO: However, authors also do not mention how to go about it
+    for i in range(len(keys_immutable)):
+        immutable_constraint_matrix1, immutable_constraint_matrix2 = build_constraints(
+            data, i, keys_immutable
+        )
+
+    # POSITIVE PREDICTIONS
+    y_predicted = model.predict_proba(data.values)
+    y_predicted = np.argmax(y_predicted, axis=1)
+    y_positive_indeces = np.where(y_predicted == 1)
+
+    # TODO: replace counterfactual search by function, which takes 'mode' as an argument
+    candidate_counterfactuals_star = []
```

Comment:
Lets move this variable decleration close to where its used. E.g. right above the for loop. 

Also nitpick: `candidate_counterfactuals` **_star** sound very paper like, in my opinion that doesn't need to be reflected in the source code. Even more since there is no non-star variables. So for me renaming to `candidate_counterfactuals` would be clearer